exorcist/tests/test_taskstatusdb.py

```diff
@@ -169,8 +189,72 @@ def test_get_all_tasks(self, request, fixture):
         expected = {
             'fresh_db': set(),
             'loaded_db': {('foo', TaskStatus.AVAILABLE.value, None, 0, 3),
-                          ('bar', TaskStatus.BLOCKED.value, None, 0, 3)},
+                ('bar', TaskStatus.BLOCKED.value, None, 0, 3)},
         }[fixture]
         db = request.getfixturevalue(fixture)
         tasks = set(db.get_all_tasks())
         assert tasks == expected
+
+    def test_add_task(self, fresh_db):
+        # task without prerequisites
+        fresh_db.add_task("foo", requirements=[], max_tries=3)
+        expected_foo_task = ("foo", TaskStatus.AVAILABLE.value, None, 0, 3)
+        tasks, deps = get_tasks_and_deps(fresh_db)
+        assert set(tasks) == {expected_foo_task}
+        assert set(deps) == set()
+
+        # task with prerequisites
+        fresh_db.add_task("bar", requirements=['foo'], max_tries=3)
+        tasks, deps = get_tasks_and_deps(fresh_db)
+        assert set(tasks) == {expected_foo_task,
+                              ("bar", TaskStatus.BLOCKED.value, None, 0, 3)}
+        assert set(deps) == {("foo", "bar")}
+
+    def test_add_task_before_requirements(self, fresh_db):
+        with pytest.raises(sqla.exc.IntegrityError, match="FOREIGN KEY"):
+            fresh_db.add_task("bar", requirements=["foo"], max_tries=3)
+
+        # check that task insertion got rolled back
+        self.assert_is_fresh_db(fresh_db)
+
+    def test_add_task_network(self, fresh_db, diamond_taskid_network):
+        fresh_db.add_task_network(diamond_taskid_network, max_tries=3)
+        tasks, deps = get_tasks_and_deps(fresh_db)
+        expected_tasks = {
+            ("A", TaskStatus.AVAILABLE.value, None, 0, 3),
+            ("B", TaskStatus.BLOCKED.value, None, 0, 3),
+            ("C", TaskStatus.BLOCKED.value, None, 0, 3),
+            ("D", TaskStatus.BLOCKED.value, None, 0, 3),
+        }
+        expected_deps = {("A", "B"), ("A", "C"), ("B", "D"), ("C", "D")}
+        assert set(tasks) == expected_tasks
+        assert set(deps) == expected_deps
+        assert len(tasks) == len(expected_tasks)
+        assert len(deps) == len(expected_deps)
+
+    def test_task_row_update_statement(self, loaded_db):
+        # TODO: I'm going to do this in a future PR
+        ...
+
+    def test_check_out_task(self, loaded_db):
+        taskid = loaded_db.check_out_task()
+        assert taskid == "foo"
+
```

Comment:
could we also test that the database doesn't repeatedly hand out the same task on multiple calls?  So we put in two tasks with no deps, and check that we don't get given 'foo' twice.

then similarly if we had foo, bar, baz where baz depends on foo and bar, can we check that only foo and bar get handed out?